asrtoolkit/alignment/align.py

```diff
@@ -0,0 +1,181 @@
+      using WhitespaceTokenizer
+    :param: lattice: List[Dict];
+      inner dicts contain token-level information for each token in input text
+    :param: token_key: key in Dicts containing tokens in text
+    :return: spacy Doc
+      (1) text: ' '.join(dict['token'])
+      (2) metadata: doc.user_data = lattice (word-level dicts)
+        - Scribe output: word-level time offsets
+        - reference transcript: speaker id and gender
+
+    * Note: spacy model must be loaded and bound to NLP object
+        NLP = spacy.load('en_core_web_sm')
+    """
+    NLP.make_doc = WhitespaceTokenizer(NLP)
+    word_list = list(pluck(token_key, lattice))
+    doc = NLP(word_list)
+    doc.user_data = lattice
+    assert len(doc) == len(word_list)  # 1:1 token:meta_dict
+    return doc
+
+
+def select_extractors(use_unigrams=False):
+    """
+    Extractors For Alignment
+    :return: List of Extractors objects to use for text-text alignment
+    note: ngram extractors below filter out stopwords and number words/symbols
+    """
+    noun_chunk_extractor = Extractor(
+        lambda doc: list(filter(lambda x: len(x) > 3, list(noun_chunks(doc)))))
+    tetragram_extractor = Extractor(
+        lambda doc: list(ngrams(doc, 4, filter_stops=True, filter_nums=True)))
+    trigram_extractor = Extractor(
+        lambda doc: list(ngrams(doc, 3, filter_stops=True, filter_nums=True)))
+    bigram_extractor = Extractor(lambda doc: list(
+        ngrams(doc, 2, filter_stops=False, filter_nums=False)))
+    unigram_extractor = Extractor(lambda doc: list(
+        ngrams(doc, 1, filter_stops=False, filter_nums=False)))
+
+    extractor_list = [
+        noun_chunk_extractor,
+        tetragram_extractor,
+        trigram_extractor,
+        bigram_extractor,
+    ]
+
+    if use_unigrams:
+        extractor_list.append(unigram_extractor)
+
+    return extractor_list
+
+
+def lattices_to_aligned_doc(time_annotated_tokens, reference_tokens,
+                            extractors, num_extractions):
+    """
+    From two Lists of Word-Dicts, returns AlignedDoc (already aligned)
+    :param time_annotated_tokens: List[Dict]; dicts with token-specific data
+      including, at minimum, token start time and duration
+    :param reference_tokens: List[Dict]; dicts with token-specific data
+      including speaker(id) and gender
+    :return: AlignedDoc object; contains
+      end-product of text-text alignment and
+      methods to retrieve matches with and without token-level metadata
+    """
+    # Create Spacy Docs from List of Tokens + Metadata
+    hyp = init_spacy_document(lattice=time_annotated_tokens,
+                              token_key="gk_token")
+    ref = init_spacy_document(lattice=reference_tokens, token_key="token")
+    return AlignedDoc(hyp, ref, extractors, num_extractions)
+
+
+def get_segments_from_alignments(
+    aligned_token_lattice,
+    reference_doc,
+    token_idx="token_idx",
+    max_duration=15,
+):
```

Comment:
Declaration formatting not consistent with function `lattices_to_aligned_doc`.

Reply:
formatting standardized locally - will push shortly